tmk/cpp/bin/tmk-clusterize.cpp

```diff
@@ -224,45 +231,75 @@ void snowballClusterize(
     float c1,
     float c2,
     bool level1Only,
+    bool verbose,
     std::map<std::string, std::set<std::string>>& equivalenceClasses,
     char* argv0) {
   std::map<std::string, std::set<std::string>> adjacencyMatrix;
 
   // COMPUTE THE ADJACENCY MATRIX
-  for (const auto& it1 : metadataToFeatures) {
-    const std::string& filename1 = it1.first;
-    const std::shared_ptr<TMKFeatureVectors> pfv1 = it1.second;
-    // printf("... %s\n", filename1.c_str());
-    for (const auto& it2 : metadataToFeatures) {
-      const std::string& filename2 = it2.first;
-      const std::shared_ptr<TMKFeatureVectors> pfv2 = it2.second;
-      // The adjacency matrix is symmetric. So write both sides of the
-      // diagonal, but do the math only one per pair.
-      if (filename1 <= filename2) {
-        if (!TMKFeatureVectors::areCompatible(*pfv1, *pfv2)) {
-          fprintf(
-              stderr,
-              "%s: immiscible provenances:\n%s\n%s\n",
-              argv0,
-              filename1.c_str(),
-              filename2.c_str());
-          exit(1);
-        }
-        float s1 = TMKFeatureVectors::computeLevel1Score(*pfv1, *pfv2);
-        if (s1 >= c1) {
-          if (level1Only) {
-            adjacencyMatrix[filename1].insert(filename2);
-            adjacencyMatrix[filename2].insert(filename1);
-          } else {
-            float s2 = TMKFeatureVectors::computeLevel2Score(*pfv1, *pfv2);
-            if (s2 >= c2) {
-              adjacencyMatrix[filename1].insert(filename2);
-              adjacencyMatrix[filename2].insert(filename1);
+
+  std::chrono::time_point<std::chrono::system_clock> startAdjacencyMatrix =
+      std::chrono::system_clock::now();
+
+  if (verbose) {
+    printf("\n");
+    printf("CALCULATING THE ADJACENCY MATRIX\n");
+  }
+
+  #pragma omp parallel
+  #pragma omp single nowait
+  {
+    for (const auto& it1 : metadataToFeatures) {
```

Comment:
Dumb question - why not parallel for?

https://stackoverflow.com/questions/1448318/omp-parallel-vs-omp-parallel-for

Reply:
Using `parallel for` was actually my first attempt, but it gave a bunch of errors that I couldn't understand. I think it doesn't like the use of map in parallel for. There is a way to use the collapse syntax with parallel for, e.g. https://stackoverflow.com/a/28483812/4626770 that seems very applicable if that could be sorted out.